Metadata and runtime support for suppressible protocol requirements

DougGregor opened the pull request:
Introduce metadata and runtime support for describing conformances to
"suppressible" protocols such as `Copyable`. The metadata changes occur
in several different places:

* Context descriptors gain a flag bit to indicate when the type itself has
  suppressed one or more suppressible protocols (e.g., it is `~Copyable`).
  When the bit is set, the context will have a trailing
  `SuppressibleProtocolSet`, a 16-bit bitfield that records one bit for
  each suppressed protocol. Types with no suppressed conformances will
  leave the bit unset (so the metadata is unchanged), and older runtimes
  don't look at the bit, so they will ignore the extra data.
* Generic context descriptors gain a flag bit to indicate when the type
  has conditional conformances to suppressible protocols. When set,
  there will be trailing metadata containing another
  `SuppressibleProtocolSet` (a subset of the one in the main context
  descriptor) indicating which suppressible protocols have conditional
  conformances, followed by the actual lists of generic requirements
  for each of the conditional conformances. Again, if there are no
  conditional conformances to suppressible protocols, the bit won't be
  set. Old runtimes ignore the bit and any trailing metadata.
* Generic requirements get a new "kind", which provides an ignored
  protocol set (another `SuppressibleProtocolSet`) stating which
  suppressible protocols should *not* be checked for the subject type
  of the generic requirement. For example, this encodes a requirement
  like `T: ~Copyable`. These generic requirements can occur anywhere
  that there is a generic requirement list, e.g., conditional
  conformances and extended existentials. Older runtimes handle unknown
  generic requirement kinds by stating that the requirement isn't
  satisfied.

Extend the runtime to perform checking of the suppressible
conformances on generic arguments as part of checking generic
requirements. This checking follows the defaults of the language, which
is that every generic argument must conform to each of the suppressible
protocols unless there is an explicit generic requirement that states
which suppressible protocols to ignore. Thus, a generic parameter list
`<T, Y where T: ~Escapable>` will check that `T` is `Copyable` but
not that it is `Escapable`, and check that `U` is both `Copyable` and
`Escapable`. To implement this, we collect the ignored protocol sets
from these suppressed requirements while processing the generic
requirements, then check all of the generic arguments against any
conformances not suppressed.

Answering the actual question "does `X` conform to `Copyable`?" (for
any suppressible protocol) looks at the context descriptor metadata to
answer the question, e.g.,

1. If there is no "suppressed protocol set", then the type conforms.
This covers types that haven't suppressed any conformances, including
all types that predate noncopyable generics.
2. If the suppressed protocol set doesn't contain `Copyable`, then the
type conforms.
3. If the type is generic and has a conditional conformance to
`Copyable`, evaluate the generic requirements for that conditional
conformance to answer whether it conforms.

The procedure above handles the bits of a `SuppressibleProtocolSet`
opaquely, with no mapping down to specific protocols. Therefore, the
same implementation will work even with future suppressible protocols,
including back deployment.

The end result of this is that we can dynamically evaluate conditional
conformances to protocols that depend on conformances to suppressible
protocols.

Implements rdar://123466649.

DougGregor committed:
Move "suppressible protocol" information into a separate .def file

We need to track suppressible protocols as part of metadata in the ABI.

DougGregor committed:
Remove never-used "version" field from `ContextDescriptorFlags`.

DougGregor committed:
Metadata and runtime support for suppressible protocol requirements

Introduce metadata and runtime support for describing conformances to
"suppressible" protocols such as `Copyable`. The metadata changes occur
in several different places:

* Context descriptors gain a flag bit to indicate when the type itself has
  suppressed one or more suppressible protocols (e.g., it is `~Copyable`).
  When the bit is set, the context will have a trailing
  `SuppressibleProtocolSet`, a 16-bit bitfield that records one bit for
  each suppressed protocol. Types with no suppressed conformances will
  leave the bit unset (so the metadata is unchanged), and older runtimes
  don't look at the bit, so they will ignore the extra data.
* Generic context descriptors gain a flag bit to indicate when the type
  has conditional conformances to suppressible protocols. When set,
  there will be trailing metadata containing another
  `SuppressibleProtocolSet` (a subset of the one in the main context
  descriptor) indicating which suppressible protocols have conditional
  conformances, followed by the actual lists of generic requirements
  for each of the conditional conformances. Again, if there are no
  conditional conformances to suppressible protocols, the bit won't be
  set. Old runtimes ignore the bit and any trailing metadata.
* Generic requirements get a new "kind", which provides an ignored
  protocol set (another `SuppressibleProtocolSet`) stating which
  suppressible protocols should *not* be checked for the subject type
  of the generic requirement. For example, this encodes a requirement
  like `T: ~Copyable`. These generic requirements can occur anywhere
  that there is a generic requirement list, e.g., conditional
  conformances and extended existentials. Older runtimes handle unknown
  generic requirement kinds by stating that the requirement isn't
  satisfied.

Extend the runtime to perform checking of the suppressible
conformances on generic arguments as part of checking generic
requirements. This checking follows the defaults of the language, which
is that every generic argument must conform to each of the suppressible
protocols unless there is an explicit generic requirement that states
which suppressible protocols to ignore. Thus, a generic parameter list
`<T, Y where T: ~Escapable>` will check that `T` is `Copyable` but
not that it is `Escapable`, and check that `U` is both `Copyable` and
`Escapable`. To implement this, we collect the ignored protocol sets
from these suppressed requirements while processing the generic
requirements, then check all of the generic arguments against any
conformances not suppressed.

Answering the actual question "does `X` conform to `Copyable`?" (for
any suppressible protocol) looks at the context descriptor metadata to
answer the question, e.g.,

1. If there is no "suppressed protocol set", then the type conforms.
This covers types that haven't suppressed any conformances, including
all types that predate noncopyable generics.
2. If the suppressed protocol set doesn't contain `Copyable`, then the
type conforms.
3. If the type is generic and has a conditional conformance to
`Copyable`, evaluate the generic requirements for that conditional
conformance to answer whether it conforms.

The procedure above handles the bits of a `SuppressibleProtocolSet`
opaquely, with no mapping down to specific protocols. Therefore, the
same implementation will work even with future suppressible protocols,
including back deployment.

The end result of this is that we can dynamically evaluate conditional
conformances to protocols that depend on conformances to suppressible
protocols.

Implements rdar://123466649.

mikeash commented on include/swift/ABI/SuppressibleProtocols.def:
```diff
@@ -0,0 +1,33 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
```
Can bump this to 2024.

mikeash commented on include/swift/ABI/SuppressibleProtocols.h:
```diff
@@ -0,0 +1,224 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
```
Likewise 2024.

mikeash commented on include/swift/ABI/SuppressibleProtocols.h:
```diff
@@ -0,0 +1,224 @@
+  constexpr bool contains##Name() const {            \
+    return contains(SuppressibleProtocolKind::Name); \
+  }
+#include "swift/ABI/SuppressibleProtocols.def"
```
A static assert that the value for `Bit` fits into `StorageType` might be nice. Unlikely we'd ever hit it, but would save some pain in the off chance we did.

DougGregor replied:
Sure. I'll bury it in a C++ file somewhere

mikeash reviewed (approved):
Looks good!

I'm sure you've thought about this so this is just me being paranoid and double-checking: we're pretty sure we'll never need more than 16 of these? I can't think of anything close to 14 more that we'd ever want to add, but, paranoid.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test

DougGregor commented:
I still have more to do, but this is passing all tests locally (finally).

DougGregor commented:
> I'm sure you've thought about this so this is just me being paranoid and double-checking: we're pretty sure we'll never need more than 16 of these? I can't think of anything close to 14 more that we'd ever want to add, but, paranoid.

Everyone seems to think we'll be okay with 16. If not, we'll have a non-back-deployable issue with the 17th.

DougGregor committed:
Extend runtime checking for suppressible protocols

Add more runtime support for checking suppressible protocol requirements:
* Parameter packs now check all of the arguments appropriately
* Most structural types now implement checking (these are hard to test).

DougGregor committed:
[Runtime] Check function types against suppressible protocols

Form a set of suppressed protocols for a function type based on
the extended flags (where future compilers can start recording
suppressible protocols) and the existing "noescape" bit. Compare
that against the "ignored" suppressible protocol requirements, as we
do for other types.

This involves a behavior change if any client has managed to evade the
static checking for noescape function types, but it's unlikely that
existing code has done so (and it was unsafe anyway).

DougGregor commented:
Added support for variadic generics and structural types, including "noescape" function types (based on the existing bits).

DougGregor commented:
@swift-ci please test

al45tair reviewed (approved):
Spotted a few typos. Otherwise looks good.

DougGregor commented:
Help, I've been replaced by a snake who puts too many "s"'s in everything

DougGregor commented on test/Interpreter/escapable_generics_casting.swift:
```diff
@@ -0,0 +1,50 @@
+  attemptCall(Dog<Nonescapable>())
+
+  // CHECK: function types
+  print("function types")
```
It's effectively a way to resync FileCheck with that place in the source file. This test should probably switch over to the unit test machinery instead of FileCheck.

DougGregor committed:
Fix typos spotted by code review, thank you!

DougGregor commented:
@swift-ci please smoke test

DougGregor commented on include/swift/ABI/GenericContext.h:
```diff
@@ -467,7 +527,84 @@ class TrailingGenericContextObjects<TargetSelf<Runtime>,
+    // Count the number of "set" bits up to (but not including) the
+    // bit we're looking at.
+    unsigned targetBit = static_cast<uint8_t>(kind);
+    auto suppressedBits = conditionallySuppressed.rawBits();
+    unsigned priorBits = 0;
+    for (unsigned i = 0; i != targetBit; ++i) {
+      if (suppressedBits & 0x01)
+        ++priorBits;
+      suppressedBits = suppressedBits >> 1;
+    }
```
We'd probably need to pull in the header. I'd like to handle this as a follow-up, because dragging another LLVM header into the runtime is the kind of thing that will find a way to blow up on me somewhere.

mikeash commented on include/swift/ABI/Metadata.h:
```diff
@@ -4699,6 +4770,23 @@ class swift_ptrauth_struct_context_descriptor(EnumDescriptor)
+  /// that are suppressed here, but that information is recorded in the
+  /// conditional suppressed protocols of the corresponding `GenericContext`.
+  const SuppressibleProtocolSet &
+  getSuppresssedProtocols() const {
```
That's just snake case.

DougGregor commented:
@swift-ci please smoke test Windows

DougGregor committed:
[Runtime] Don't check suppressible protocols for non-key generic arguments

DougGregor commented:
@swift-ci please test

DougGregor committed:
Drop constexpr rejected by Visual C++

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
Remote more constexpr's that Visual C++ doesn't like

DougGregor commented:
@swift-ci please smoke test
